tests/pubsub/testgossipmembership.nim

```diff
@@ -0,0 +1,255 @@
+
+    await allFuturesThrowing(nodes.mapIt(allFutures(it.switch.stop())))
+
+  # Test ensuring the number of subscriptions does not exceed a set limit
+  asyncTest "subscription limit test":
+    # Given one node and a subscription limit of 10 topics
+    let
+      topicCount = 15
+      gossipSubParams = 10
+      topicNames = toSeq(mapIt(0 .. topicCount - 1, "topic" & $it))
+      numberOfNodes = 1
+      nodes = generateNodes(numberOfNodes, gossip = true)
+
+    await allFuturesThrowing(nodes.mapIt(it.switch.start()))
+
+    # When attempting to subscribe to 15 topics
+    let gossipSub = GossipSub(nodes[0])
+    gossipSub.topicsHigh = gossipSubParams
+
+    for topic in topicNames:
+      if gossipSub.topics.len < gossipSub.topicsHigh:
+        gossipSub.subscribe(
+          topic,
+          proc(topic: string, data: seq[byte]): Future[void] {.async.} =
+            discard
+          ,
+        )
+      else:
+        # Then the subscription count should not exceed the limit
+        check gossipSub.topics.len == gossipSub.topicsHigh
+
+    check gossipSub.topics.len == gossipSub.topicsHigh
+
+    await allFuturesThrowing(nodes.mapIt(allFutures(it.switch.stop())))
+
+  # Test verifying peers joining a topic using `JOIN(topic)`
+  asyncTest "handle JOIN topic and mesh is updated":
+    # Given 5 nodes and a join request to a topic
+    let
+      topic = "test-join-topic"
+      numberOfNodes = 5
+      nodes = generateNodes(numberOfNodes, gossip = true)
+
+    await allFuturesThrowing(nodes.mapIt(it.switch.start()))
+
+    # When nodes join and subscribe to the topic
+    await subscribeNodes(nodes)
+    for node in nodes:
+      node.subscribe(topic, voidTopicHandler)
+
+    await sleepAsync(2 * DURATION_TIMEOUT)
+
+    # Then each node's mesh should reflect this update
+    for node in nodes:
+      let currentGossip = GossipSub(node)
+      check currentGossip.mesh[topic].len == numberOfNodes - 1
+      check currentGossip.topics.contains(topic)
+
+    await allFuturesThrowing(nodes.mapIt(allFutures(it.switch.stop())))
+
+  # Test the behavior when multiple peers join and leave a topic simultaneously
+  asyncTest "multiple peers join and leave topic simultaneously":
+    # Given 6 nodes and a shared topic
+    let
+      numberOfNodes = 6
+      topic = "foobar"
+      nodes = generateNodes(numberOfNodes, gossip = true)
+      nodesFut = await allFinished(nodes.mapIt(it.switch.start()))
+
+    await subscribeNodes(nodes)
+    for node in nodes:
+      node.subscribe(topic, voidTopicHandler)
+
+    await sleepAsync(2 * DURATION_TIMEOUT)
+
+    # When they all join the topic, their attributes should reflect this
+    for i in 0 ..< numberOfNodes:
+      let currentGossip = GossipSub(nodes[i])
+      check currentGossip.gossipsub.hasKey(topic)
+      check currentGossip.mesh.hasKey(topic)
+      check currentGossip.topics.contains(topic)
+
+    # When peers subscribe to each other's topics
```

Comment:
The subscription is done above, on the `subscribe`. This is just a synchronous check to turn the test deterministic: Make sure all nodes are connected between themselves.